### ISSUE TITLE
GHA for 4.8.4 on release failed to produce changelog.

### ISSUE THREAD

**Issue by hazendaz** [spotbugs/spotbugs#2921]
The first build failed midway but it did push to sonatype.  I ran it again and it pushed to sonatype again but both cases it could not figure out the tag.  I pushed up the 2 commits as usual at same time.  Not sure what is going on but we need to either manually add change log or fix the issue and run again.  If running again, will need to drop the sonatype push.

**Comment by JuditKnoll** [comment]
Here is the workflow run: https://github.com/spotbugs/spotbugs/actions/runs/8589978985
As far as I can see, there are two separate problems: 
1. the release itself is not created,
1. the changelog is not created.


It looks like the commits were directly pushed to the master. 
@hazendaz Did you push the tags and the commits the same time? Maybe, that can have an impact on the outcome of the first issue.
Where was the tags created? Directly on SpotBugs/master?


What I find quite strange is, that the gha is running on a tag, 4.8.4, which seems to be on master, but the `Commit CHANGELOG.md`, the `Build on tag` and `Create Release` steps are skipped, and the `Build otherwise` step runs. This means that relevant condition (which is supposed to check, whether it runs on a tag, which is on the master branch) fails.
This is why the release in GitHub is not created, and the commit is handled as any other simple commit.
I'll look into this issue and try it out on my fork.

The other half of the problem is the creation of the changelog itself.
The `Get previous tag for the change log` step finds the current tag as the previous one.

**Comment by gtoison** [comment]
I'm not able to test the change but I think https://github.com/spotbugs/spotbugs/pull/2925 should help with this problem

**Comment by JuditKnoll** [comment]
Tested the changes of https://github.com/spotbugs/spotbugs/pull/2925 on my fork, and could find a solution to part of the second problem (see the details at the PR), but commiting the changelog file will be a problem. Currently the release process is triggered by pushing a tag on the master branch on spotbugs/spotbugs repository. On a tag push event, the github action is not on a branch, but in "detached HEAD" state, where no commits can be commited. Also, there are already at least one commit after the commit from which the release is created. 

As far as I can see if we would like to keep the `CHANGELOG.md` file, and automatically update it during the release process, we need to change the release process trigger. Also, IMO it would be really nice, if the commit updating the changelog would be in the release itself, so before the commit the release is created from, but this means that either the changelog updating commit needs to be run separately or the tag should be created by the release process. What do you guys think?

**Comment by gtoison** [comment]
Sorry for causing a mess with this attempt to automate the changelog...
Does anyone know (or use?) JReleaser? Among its many features and integrations it does changelog management (https://jreleaser.org/guide/latest/reference/release/changelog.html)
It seems like it could make the release process simpler

**Comment by JuditKnoll** [comment]
Checked it once more, every link from the release process, and it looks like, only the release is only missing from the GitHub Release page. In this case IMO this small part should be done manually, not by retriggering the whole release process.
Probably because the `isReleaseVersion` check in [maven.gradle](https://github.com/spotbugs/spotbugs/blob/master/gradle/maven.gradle#L4):
```
ext.isReleaseVersion = !version.endsWith("SNAPSHOT") && /* for Eclipse plugin */ !version.contains("-SNAPSHOT.")
```

> The tag could be created on commit that commits changelog during the release process. Makes sense for me.

Good idea. In this case even the first commit after the release, the `prepare for the next release` commit also could be pushed by the release process, just to make sure that it is not forgotten.

> Sorry for causing a mess with this attempt to automate the changelog...

A large portion of the problems are because of the conditions I introduced, but I'm still don't get why it caused problems, when on my fork, it worked... I'm sorry for the issues. 
@gtoison Thank you for your work, which does bring this project forward 😄 

> Does anyone know (or use?) JReleaser? Among its many features and integrations it does changelog management (https://jreleaser.org/guide/latest/reference/release/changelog.html)

I don't know JReleaser. I think the changelog management could be solved by the github action you added, but the pushing the commit updating the changelog file is a bit problematic with the current release process, but it would be with any changelog management system.

**Comment by hazendaz** [comment]
Maybe we need another GHA that just publishes the change log only so if we have issues, we can just run that and have it do it in the way it does.  An on demand action that would require we provide the tags to check between might do the trick there.  Anyone want to give that a try?

**Comment by JuditKnoll** [comment]
> Maybe we need another GHA that just publishes the change log only so if we have issues, we can just run that and have it do it in the way it does. An on demand action that would require we provide the tags to check between might do the trick there. Anyone want to give that a try?

By "publishes the change log" do you mean pushes a commit updating the CHANGELOG.md file, or creates the changelog for the GitHub release? 
If the prior, then IMO it should not work with tags, if we want to keep the tag push as a trigger of the release process and include the changelog updating commit in the new release.
If the latter, then this can be done by the release process itself without drastic changes, so I don't think a separate GHA is needed for it.

**Comment by gtoison** [comment]
I think the source zip is generated on from the git tree (for the relevant tag/branch/etc.) by github, this seems to work for me https://github.com/spotbugs/spotbugs/archive/refs/tags/4.8.4.zip

**Comment by JuditKnoll** [comment]
> I think the source zip is generated on from the git tree (for the relevant tag/branch/etc.) by github, this seems to work for me https://github.com/spotbugs/spotbugs/archive/refs/tags/4.8.4.zip

I think the one generated by GitHub is a different one. The usual releases spotbugs releases have the following 8 assets:
- `eclipsePlugin.zip` - couldn't find it,
- `spotbugs-4.8.4-source.zip` - couldn't find it,
- `spotbugs-4.8.4.tgz` - got from the Nexus Sonatype repository,
- `spotbugs-4.8.4.tgz.asc` - got from the Nexus Sonatype repository,
- `spotbugs-4.8.4.zip` - got from the Nexus Sonatype repository,
- `spotbugs-4.8.4.zip.asc` - got from the Nexus Sonatype repository,
- Source code (zip) - generated by GitHub,
- Source code (tar) - generated by GitHub.

Compared the `spotbugs-4.8.3-source.zip` and the GitHub generated zip for 4.8.3 version, and there were some differences. 
If anyone knows, how to get the first two archives, feel free to add to the assets. Or should we just publish the release without these two assets?
________________________________

However, this is off topic to the original issue, which I think was fixed by https://github.com/spotbugs/spotbugs/pull/2925 (at least the changelog creation part). For the commit creation part, we need to rework the whole release process, which should have its own issue IMO.
For avoiding the problem with the GitHub release not being created in the future, I opened https://github.com/spotbugs/spotbugs/pull/2947. For solving the immediate problem with the 4.8.4 GitHub release, drafted the release.

**Comment by hazendaz** [comment]
For eclipse its using the gh-pages off the 3 eclipse repos in this project.  I've updated all these readme's to note this is done via GHA not TravisCI.  I simply linked those back to the GHA release action.

I'm not sure it matters for Eclipse to be part of the releases on github.  Its served differently.  Maybe that is useful for someone that has to do offline install.  I'm not sure how to pull from there.  However, I can confirm Eclipse does show the release from 4/7/2024.

I'm also not so sure it matters that source zips are listed there either.

While I'm sure they might be nice to have the source is in github as it is.  IMO I think you can just take draft off the release and let it be.  I suspect we are following up here quickly anyways and going to publish hopefully working 4.8.5 release that gets us past this situation.  At that point I think it would matter even less.

**Comment by hazendaz** [comment]
![image](https://github.com/spotbugs/spotbugs/assets/975267/ff6c2f4d-47e3-48ef-8b62-8d3faf685c94)

**Comment by hazendaz** [comment]
For eclipse manual download, anyone needing that can just check out the gh-pages branch https://github.com/spotbugs/eclipse/tree/gh-pages.  That is the full content of the site eclipse uses to install.  It doesn't need zipped really but we could just grab from there, zip it and add it to the content manually.

**Comment by iloveeclipse** [comment]
> I'm not sure it matters for Eclipse to be part of the releases on github. Its served differently. Maybe that is useful for someone that has to do offline install. I'm not sure how to pull from there

It matters, and yes, it is for offline install.

**Comment by JuditKnoll** [comment]
@iloveeclipse Do you know a better place to get the `eclipsePlugin.zip` to attach to the release? Or should I just download, zip and upload it from https://github.com/spotbugs/eclipse/tree/gh-pages?

**Comment by iloveeclipse** [comment]
@JuditKnoll : do you mean automated creation or for a one-time 4.8.4 fix? For later, yes, should be enough to create it manually as you've proposed, just make sure to get right commit.

**Comment by JuditKnoll** [comment]
> @JuditKnoll : do you mean automated creation or for a one-time 4.8.4 fix? For later, yes, should be enough to create it manually as you've proposed, just make sure to get right commit.

For the automated creation, I proposed https://github.com/spotbugs/spotbugs/pull/2947 which creates a draft PR for every tag, but only publishes it, if it's on the master. It's already got merged.
Added to 4.8.4 the eclipsePlugin.zip as well, getting it the newest commit from https://github.com/spotbugs/eclipse/tree/gh-pages, which is for the release of 4.8.4. Downloaded, unzipped and rezipped the inner folder, since otherwise there would have been an unnecessary inner folder.
If no one is against it, I will publish the release for 4.8.4 on the github page tomorrow the way it is now.
